### PR TITLE
Add mypy and update typings

### PR DESCRIPTION
**Pull request by sloria**
Since this package distributes types, it'd be good to ensure that they're correct. This adds mypy to the pre-commit config and updates type annotations to pass mypy's type checks.

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -114,37 +114,6 @@
 htmlhelp_basename = "marshmallow_dataclassdoc"
 
 
-# -- Options for LaTeX output ------------------------------------------------
-
-latex_elements = {
```
Comment (sloria) [inline docs/conf.py]: mypy was asking to type this, but it looks like it's unused anyway.

**Comment by lovasoa** [comment]
Great !

@karlb, you finally get correct types for this library :)

File: marshmallow_dataclass/__init__.py [hunk 2]
```diff
@@ -430,8 +460,10 @@ def _get_field_default(field: dataclasses.Field):
     >>> _get_field_default(dataclasses.field())
     <marshmallow.missing>
     """
-    if field.default_factory is not dataclasses.MISSING:
-        return field.default_factory
+    # Remove `type: ignore` when https://github.com/python/typeshed/issues/3246 is fixed
```
Comment (lovasoa) [inline marshmallow_dataclass/__init__.py]: Maybe this is the issue that should be linked instead:
https://github.com/python/mypy/issues/6910

**Comment by karlb** [comment]
Cool, thanks!

File: marshmallow_dataclass/__init__.py (hunk 2 above)
Comment (sloria) [reply marshmallow_dataclass/__init__.py]: Woops meant to update that. Thanks. Fixed.

**Comment by sloria** [comment]
@lovasoa Shall I merge my own PRs after you approve or would you like to do that?

**Comment by lovasoa** [comment]
@sloria, you can merge your own PRs !

**Comment by sloria** [comment]
OK, I'll squash and merge after your approval from now on.

I'll be responsive most days, but of course, no need to block on me--feel free to merge when it's practical.

**Comment by lovasoa** [comment]
The same for me. Feel free to merge things if I'm not responsive.

**Pull request merged**